// @libp2p/record/dist/src/record.js
import { decodeMessage, encodeMessage, message, streamMessage } from 'protons-runtime';
import { alloc as uint8ArrayAlloc } from 'uint8arrays/alloc';
export var Record;
(function (Record) {
    let _codec;
    Record.codec = () => {
        if (_codec == null) {
            _codec = message((obj, w, opts = {}) => {
                if (opts.lengthDelimited !== false) {
                    w.fork();
                }
                if ((obj.key != null && obj.key.byteLength > 0)) {
                    w.uint32(10);
                    w.bytes(obj.key);
                }
                if ((obj.value != null && obj.value.byteLength > 0)) {
                    w.uint32(18);
                    w.bytes(obj.value);
                }
                if ((obj.timeReceived != null && obj.timeReceived !== '')) {
                    w.uint32(42);
                    w.string(obj.timeReceived);
                }
                if (opts.lengthDelimited !== false) {
                    w.ldelim();
                }
            }, (reader, length, opts = {}) => {
                const obj = {
                    key: uint8ArrayAlloc(0),
                    value: uint8ArrayAlloc(0),
                    timeReceived: ''
                };
                const end = length == null ? reader.len : reader.pos + length;
                while (reader.pos < end) {
                    const tag = reader.uint32();
                    switch (tag >>> 3) {
                        case 1: {
                            obj.key = reader.bytes();
                            break;
                        }
                        case 2: {
                            obj.value = reader.bytes();
                            break;
                        }
                        case 5: {
                            obj.timeReceived = reader.string();
                            break;
                        }
                        default: {
                            reader.skipType(tag & 7);
                            break;
                        }
                    }
                }
                return obj;
            }, function* (reader, length, prefix, opts = {}) {
                const end = length == null ? reader.len : reader.pos + length;
                while (reader.pos < end) {
                    const tag = reader.uint32();
                    switch (tag >>> 3) {
                        case 1: {
                            yield {
                                field: `${prefix}.key`,
                                value: reader.bytes()
                            };
                            break;
                        }
                        case 2: {
                            yield {
                                field: `${prefix}.value`,
                                value: reader.bytes()
                            };
                            break;
                        }
                        case 5: {
                            yield {
                                field: `${prefix}.timeReceived`,
                                value: reader.string()
                            };
                            break;
                        }
                        default: {
                            reader.skipType(tag & 7);
                            break;
                        }
                    }
                }
            });
        }
        return _codec;
    };
    function encode(obj) {
        return encodeMessage(obj, Record.codec());
    }
    Record.encode = encode;
    function decode(buf, opts) {
        return decodeMessage(buf, Record.codec(), opts);
    }
    Record.decode = decode;
    function stream(buf, opts) {
        return streamMessage(buf, Record.codec(), opts);
    }
    Record.stream = stream;
})(Record || (Record = {}));
