// @libp2p/kad-dht/dist/src/message/dht.js
import { decodeMessage, encodeMessage, enumeration, MaxLengthError, message, streamMessage } from 'protons-runtime';
import { alloc as uint8ArrayAlloc } from 'uint8arrays/alloc';
export var Record;
(function (Record) {
    let _codec;
    Record.codec = () => {
        if (_codec == null) {
            _codec = message((obj, w, opts = {}) => {
                if (opts.lengthDelimited !== false) {
                    w.fork();
                }
                if (obj.key != null) {
                    w.uint32(10);
                    w.bytes(obj.key);
                }
                if (obj.value != null) {
                    w.uint32(18);
                    w.bytes(obj.value);
                }
                if (obj.author != null) {
                    w.uint32(26);
                    w.bytes(obj.author);
                }
                if (obj.signature != null) {
                    w.uint32(34);
                    w.bytes(obj.signature);
                }
                if (obj.timeReceived != null) {
                    w.uint32(42);
                    w.string(obj.timeReceived);
                }
                if (opts.lengthDelimited !== false) {
                    w.ldelim();
                }
            }, (reader, length, opts = {}) => {
                const obj = {};
                const end = length == null ? reader.len : reader.pos + length;
                while (reader.pos < end) {
                    const tag = reader.uint32();
                    switch (tag >>> 3) {
                        case 1: {
                            obj.key = reader.bytes();
                            break;
                        }
                        case 2: {
                            obj.value = reader.bytes();
                            break;
                        }
                        case 3: {
                            obj.author = reader.bytes();
                            break;
                        }
                        case 4: {
                            obj.signature = reader.bytes();
                            break;
                        }
                        case 5: {
                            obj.timeReceived = reader.string();
                            break;
                        }
                        default: {
                            reader.skipType(tag & 7);
                            break;
                        }
                    }
                }
                return obj;
            }, function* (reader, length, prefix, opts = {}) {
                const end = length == null ? reader.len : reader.pos + length;
                while (reader.pos < end) {
                    const tag = reader.uint32();
                    switch (tag >>> 3) {
                        case 1: {
                            yield {
                                field: `${prefix}.key`,
                                value: reader.bytes()
                            };
                            break;
                        }
                        case 2: {
                            yield {
                                field: `${prefix}.value`,
                                value: reader.bytes()
                            };
                            break;
                        }
                        case 3: {
                            yield {
                                field: `${prefix}.author`,
                                value: reader.bytes()
                            };
                            break;
                        }
                        case 4: {
                            yield {
                                field: `${prefix}.signature`,
                                value: reader.bytes()
                            };
                            break;
                        }
                        case 5: {
                            yield {
                                field: `${prefix}.timeReceived`,
                                value: reader.string()
                            };
                            break;
                        }
                        default: {
                            reader.skipType(tag & 7);
                            break;
                        }
                    }
                }
            });
        }
        return _codec;
    };
    function encode(obj) {
        return encodeMessage(obj, Record.codec());
    }
    Record.encode = encode;
    function decode(buf, opts) {
        return decodeMessage(buf, Record.codec(), opts);
    }
    Record.decode = decode;
    function stream(buf, opts) {
        return streamMessage(buf, Record.codec(), opts);
    }
    Record.stream = stream;
})(Record || (Record = {}));
export var MessageType;
(function (MessageType) {
    MessageType["PUT_VALUE"] = "PUT_VALUE";
    MessageType["GET_VALUE"] = "GET_VALUE";
    MessageType["ADD_PROVIDER"] = "ADD_PROVIDER";
    MessageType["GET_PROVIDERS"] = "GET_PROVIDERS";
    MessageType["FIND_NODE"] = "FIND_NODE";
    MessageType["PING"] = "PING";
})(MessageType || (MessageType = {}));
var __MessageTypeValues;
(function (__MessageTypeValues) {
    __MessageTypeValues[__MessageTypeValues["PUT_VALUE"] = 0] = "PUT_VALUE";
    __MessageTypeValues[__MessageTypeValues["GET_VALUE"] = 1] = "GET_VALUE";
    __MessageTypeValues[__MessageTypeValues["ADD_PROVIDER"] = 2] = "ADD_PROVIDER";
    __MessageTypeValues[__MessageTypeValues["GET_PROVIDERS"] = 3] = "GET_PROVIDERS";
    __MessageTypeValues[__MessageTypeValues["FIND_NODE"] = 4] = "FIND_NODE";
    __MessageTypeValues[__MessageTypeValues["PING"] = 5] = "PING";
})(__MessageTypeValues || (__MessageTypeValues = {}));
(function (MessageType) {
    MessageType.codec = () => {
        return enumeration(__MessageTypeValues);
    };
})(MessageType || (MessageType = {}));
export var ConnectionType;
(function (ConnectionType) {
    ConnectionType["NOT_CONNECTED"] = "NOT_CONNECTED";
    ConnectionType["CONNECTED"] = "CONNECTED";
    ConnectionType["CAN_CONNECT"] = "CAN_CONNECT";
    ConnectionType["CANNOT_CONNECT"] = "CANNOT_CONNECT";
})(ConnectionType || (ConnectionType = {}));
var __ConnectionTypeValues;
(function (__ConnectionTypeValues) {
    __ConnectionTypeValues[__ConnectionTypeValues["NOT_CONNECTED"] = 0] = "NOT_CONNECTED";
    __ConnectionTypeValues[__ConnectionTypeValues["CONNECTED"] = 1] = "CONNECTED";
    __ConnectionTypeValues[__ConnectionTypeValues["CAN_CONNECT"] = 2] = "CAN_CONNECT";
    __ConnectionTypeValues[__ConnectionTypeValues["CANNOT_CONNECT"] = 3] = "CANNOT_CONNECT";
})(__ConnectionTypeValues || (__ConnectionTypeValues = {}));
(function (ConnectionType) {
    ConnectionType.codec = () => {
        return enumeration(__ConnectionTypeValues);
    };
})(ConnectionType || (ConnectionType = {}));
export var PeerInfo;
(function (PeerInfo) {
    let _codec;
    PeerInfo.codec = () => {
        if (_codec == null) {
            _codec = message((obj, w, opts = {}) => {
                if (opts.lengthDelimited !== false) {
                    w.fork();
                }
                if ((obj.id != null && obj.id.byteLength > 0)) {
                    w.uint32(10);
                    w.bytes(obj.id);
                }
                if (obj.multiaddrs != null && obj.multiaddrs.length > 0) {
                    for (const value of obj.multiaddrs) {
                        w.uint32(18);
                        w.bytes(value);
                    }
                }
                if (obj.connection != null) {
                    w.uint32(24);
                    ConnectionType.codec().encode(obj.connection, w);
                }
                if (opts.lengthDelimited !== false) {
                    w.ldelim();
                }
            }, (reader, length, opts = {}) => {
                const obj = {
                    id: uint8ArrayAlloc(0),
                    multiaddrs: []
                };
                const end = length == null ? reader.len : reader.pos + length;
                while (reader.pos < end) {
                    const tag = reader.uint32();
                    switch (tag >>> 3) {
                        case 1: {
                            obj.id = reader.bytes();
                            break;
                        }
                        case 2: {
                            if (opts.limits?.multiaddrs != null && obj.multiaddrs.length === opts.limits.multiaddrs) {
                                throw new MaxLengthError('Decode error - repeated field "multiaddrs" had too many elements');
                            }
                            obj.multiaddrs.push(reader.bytes());
                            break;
                        }
                        case 3: {
                            obj.connection = ConnectionType.codec().decode(reader);
                            break;
                        }
                        default: {
                            reader.skipType(tag & 7);
                            break;
                        }
                    }
                }
                return obj;
            }, function* (reader, length, prefix, opts = {}) {
                const obj = {
                    multiaddrs: 0
                };
                const end = length == null ? reader.len : reader.pos + length;
                while (reader.pos < end) {
                    const tag = reader.uint32();
                    switch (tag >>> 3) {
                        case 1: {
                            yield {
                                field: `${prefix}.id`,
                                value: reader.bytes()
                            };
                            break;
                        }
                        case 2: {
                            if (opts.limits?.multiaddrs != null && obj.multiaddrs === opts.limits.multiaddrs) {
                                throw new MaxLengthError('Streaming decode error - repeated field "multiaddrs" had too many elements');
                            }
                            yield {
                                field: `${prefix}.multiaddrs[]`,
                                index: obj.multiaddrs,
                                value: reader.bytes()
                            };
                            obj.multiaddrs++;
                            break;
                        }
                        case 3: {
                            yield {
                                field: `${prefix}.connection`,
                                value: ConnectionType.codec().decode(reader)
                            };
                            break;
                        }
                        default: {
                            reader.skipType(tag & 7);
                            break;
                        }
                    }
                }
            });
        }
        return _codec;
    };
    function encode(obj) {
        return encodeMessage(obj, PeerInfo.codec());
    }
    PeerInfo.encode = encode;
    function decode(buf, opts) {
        return decodeMessage(buf, PeerInfo.codec(), opts);
    }
    PeerInfo.decode = decode;
    function stream(buf, opts) {
        return streamMessage(buf, PeerInfo.codec(), opts);
    }
    PeerInfo.stream = stream;
})(PeerInfo || (PeerInfo = {}));
export var Message;
(function (Message) {
    let _codec;
    Message.codec = () => {
        if (_codec == null) {
            _codec = message((obj, w, opts = {}) => {
                if (opts.lengthDelimited !== false) {
                    w.fork();
                }
                if (obj.type != null && __MessageTypeValues[obj.type] !== 0) {
                    w.uint32(8);
                    MessageType.codec().encode(obj.type, w);
                }
                if (obj.clusterLevel != null) {
                    w.uint32(80);
                    w.int32(obj.clusterLevel);
                }
                if (obj.key != null) {
                    w.uint32(18);
                    w.bytes(obj.key);
                }
                if (obj.record != null) {
                    w.uint32(26);
                    w.bytes(obj.record);
                }
                if (obj.closer != null && obj.closer.length > 0) {
                    for (const value of obj.closer) {
                        w.uint32(66);
                        PeerInfo.codec().encode(value, w);
                    }
                }
                if (obj.providers != null && obj.providers.length > 0) {
                    for (const value of obj.providers) {
                        w.uint32(74);
                        PeerInfo.codec().encode(value, w);
                    }
                }
                if (opts.lengthDelimited !== false) {
                    w.ldelim();
                }
            }, (reader, length, opts = {}) => {
                const obj = {
                    type: MessageType.PUT_VALUE,
                    closer: [],
                    providers: []
                };
                const end = length == null ? reader.len : reader.pos + length;
                while (reader.pos < end) {
                    const tag = reader.uint32();
                    switch (tag >>> 3) {
                        case 1: {
                            obj.type = MessageType.codec().decode(reader);
                            break;
                        }
                        case 10: {
                            obj.clusterLevel = reader.int32();
                            break;
                        }
                        case 2: {
                            obj.key = reader.bytes();
                            break;
                        }
                        case 3: {
                            obj.record = reader.bytes();
                            break;
                        }
                        case 8: {
                            if (opts.limits?.closer != null && obj.closer.length === opts.limits.closer) {
                                throw new MaxLengthError('Decode error - repeated field "closer" had too many elements');
                            }
                            obj.closer.push(PeerInfo.codec().decode(reader, reader.uint32(), {
                                limits: opts.limits?.closer$
                            }));
                            break;
                        }
                        case 9: {
                            if (opts.limits?.providers != null && obj.providers.length === opts.limits.providers) {
                                throw new MaxLengthError('Decode error - repeated field "providers" had too many elements');
                            }
                            obj.providers.push(PeerInfo.codec().decode(reader, reader.uint32(), {
                                limits: opts.limits?.providers$
                            }));
                            break;
                        }
                        default: {
                            reader.skipType(tag & 7);
                            break;
                        }
                    }
                }
                return obj;
            }, function* (reader, length, prefix, opts = {}) {
                const obj = {
                    closer: 0,
                    providers: 0
                };
                const end = length == null ? reader.len : reader.pos + length;
                while (reader.pos < end) {
                    const tag = reader.uint32();
                    switch (tag >>> 3) {
                        case 1: {
                            yield {
                                field: `${prefix}.type`,
                                value: MessageType.codec().decode(reader)
                            };
                            break;
                        }
                        case 10: {
                            yield {
                                field: `${prefix}.clusterLevel`,
                                value: reader.int32()
                            };
                            break;
                        }
                        case 2: {
                            yield {
                                field: `${prefix}.key`,
                                value: reader.bytes()
                            };
                            break;
                        }
                        case 3: {
                            yield {
                                field: `${prefix}.record`,
                                value: reader.bytes()
                            };
                            break;
                        }
                        case 8: {
                            if (opts.limits?.closer != null && obj.closer === opts.limits.closer) {
                                throw new MaxLengthError('Streaming decode error - repeated field "closer" had too many elements');
                            }
                            for (const evt of PeerInfo.codec().stream(reader, reader.uint32(), `${prefix}.closer[]`, {
                                limits: opts.limits?.closer$
                            })) {
                                yield {
                                    ...evt,
                                    index: obj.closer
                                };
                            }
                            obj.closer++;
                            break;
                        }
                        case 9: {
                            if (opts.limits?.providers != null && obj.providers === opts.limits.providers) {
                                throw new MaxLengthError('Streaming decode error - repeated field "providers" had too many elements');
                            }
                            for (const evt of PeerInfo.codec().stream(reader, reader.uint32(), `${prefix}.providers[]`, {
                                limits: opts.limits?.providers$
                            })) {
                                yield {
                                    ...evt,
                                    index: obj.providers
                                };
                            }
                            obj.providers++;
                            break;
                        }
                        default: {
                            reader.skipType(tag & 7);
                            break;
                        }
                    }
                }
            });
        }
        return _codec;
    };
    function encode(obj) {
        return encodeMessage(obj, Message.codec());
    }
    Message.encode = encode;
    function decode(buf, opts) {
        return decodeMessage(buf, Message.codec(), opts);
    }
    Message.decode = decode;
    function stream(buf, opts) {
        return streamMessage(buf, Message.codec(), opts);
    }
    Message.stream = stream;
})(Message || (Message = {}));
